examples/Cassie/cassie_rbt_state_estimator.cc

```diff
@@ -816,32 +848,164 @@ void CassieRbtStateEstimator::contactEstimation(
       filtered_residual_right;
   }
 
+  // Record previous velocity (used in acceleration residual)
+  discrete_state->get_mutable_vector(
+      previous_velocity_idx_).get_mutable_value() << output.GetVelocities();
+
+  // Ground truth of contact
+  /*
+  RigidBodyTree<double>* tree_with_ground = nullptr;
+  tree_with_ground = const_cast<RigidBodyTree<double>*>(&tree_);
+  std::vector<drake::multibody::collision::PointPair<double>> pairs;
+  pairs = tree_with_ground->ComputeMaximumDepthCollisionPoints(cache,
+      true, false);
+  int gtl = 0;
+  int gtr = 0;
+  for (const auto& pair: pairs) {
+    if (pair.distance < 0.0) {
+      if (pair.elementA->get_body()->get_body_index() == 18) {
+        gtl += 1;
+      }
+      if (pair.elementA->get_body()->get_body_index() == 20) {
+        gtr += 1;
+      }
+    }
+  }
+  */
+}
+
+
+/// EstimateContactForEkf(). Conservative estimation.
+/// EKF is updated based on the assumption of stance foot being stationary.
+/// The estimated state would get very inaccurate if the stance foot is moving.
+///
+/// Input:
+///  - OutputVector `output` containing the positions, velocities and
+///    actuator torques of the robot
+/// Ouput: left contact `left_contact` and right contact `right_contact` that
+///  indicate if the corresponding foot is in contact with the ground
+///
+/// Algorithm:
+///  The contact is estimated based on
+///   1. compression in the spring
+///   2. three optimizations one for each double support, left support and
+///   right support
+///  If the compression in the left (right) heel/ankle spring is more than the
+///  set threshold, the left (right) foot is estimated to be in contact with
+///  the ground.
+///  During impact, both the legs have some non-zero acceleration and hence the
+///  optimal costs will all be high. This case is assumed to be *no* stance.
+void CassieRbtStateEstimator::EstimateContactForEkf(
+    const OutputVector<double>& output,
+    int* left_contact, int* right_contact) const {
```

Comment:
Should `optimal_cost` instead be an input to this method? This function seems to require that optimal_cost be set prior to calling it, which is always a dangerous requirement (and currently undocumented)

Reply:
I didn't add optimal_cost as an input to avoid reallocating memory, but I guess it doesn't matter too much because it's only a size-3 vector.